src/blockrelay/mempool_sync.cpp

```diff
@@ -0,0 +1,503 @@
+    // We need the number of transactions in the mempool and orphanpools but also the number
+    // in the txCommitQ that have been processed and valid, and which will be in the mempool shortly.
+    uint64_t nCommitQ = 0;
+    {
+        boost::unique_lock<boost::mutex> lock(csCommitQ);
+        nCommitQ = txCommitQ->size();
+    }
+
+    uint64_t nTxInMempool = mempool.size() + orphanpool.GetOrphanPoolSize() + nCommitQ;
+    uint64_t nMempoolMaxTxBytes = GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000;
+    uint64_t nSatoshiPerK = minRelayTxFee.GetFeePerK();
+
+    // Form SipHash keys
+    uint64_t seed = GetRand(std::numeric_limits<uint64_t>::max());
+    CDataStream stream(SER_NETWORK, PROTOCOL_VERSION);
+    stream << seed;
+    CSHA256 hasher;
+    hasher.Write((unsigned char *)&(*stream.begin()), stream.end() - stream.begin());
+    uint256 shorttxidhash;
+    hasher.Finalize(shorttxidhash.begin());
+    uint64_t shorttxidk0 = shorttxidhash.GetUint64(0);
+    uint64_t shorttxidk1 = shorttxidhash.GetUint64(1);
+
+    // Calculate how many bytes of space remain in the mempool
+    uint64_t nRemainingMempoolTxBytes = nMempoolMaxTxBytes;
+    {
+        READLOCK(mempool.cs_txmempool);
+        for (const CTxMemPoolEntry &e : mempool.mapTx)
+        {
+            nRemainingMempoolTxBytes += e.GetTx().GetTxSize();
+        }
+    }
+
+    return CMempoolSyncInfo(nTxInMempool, nRemainingMempoolTxBytes, shorttxidk0, shorttxidk1, nSatoshiPerK);
+}
+
+uint64_t NegotiateMempoolSyncVersion(CNode *pfrom)
+{
+    uint64_t peerMin = pfrom->nMempoolSyncMinVersionSupported;
+    uint64_t selfMin = mempoolSyncMinVersionSupported.Value();
+    uint64_t peerMax = pfrom->nMempoolSyncMaxVersionSupported;
+    uint64_t selfMax = mempoolSyncMaxVersionSupported.Value();
+
+    uint64_t upper = (uint64_t)std::min(peerMax, selfMax);
+    uint64_t lower = (uint64_t)std::max(peerMin, selfMin);
+
+    if (lower > upper)
+        throw std::runtime_error("Sender and receiver support incompatible mempool sync versions");
+
+    return upper;
+}
+
+CNode *SelectMempoolSyncPeer(std::vector<CNode *> vNodesCopy)
+{
+    std::vector<CNode *> vSyncableNodes;
+
+    for (auto node : vNodesCopy)
+    {
+        // Skip if mempool sync is not supported
+        if (!node->canSyncMempoolWithPeers)
+            continue;
+
+        // Skip if version cannot be negotiated
+        try
+        {
+            NegotiateMempoolSyncVersion(node);
+        }
+        catch (std::runtime_error &e)
+        {
+            continue;
+        }
+
+        CNodeStateAccessor state(nodestate, node->GetId());
+        int nCommonHeight = state->pindexLastCommonBlock ? state->pindexLastCommonBlock->nHeight : -1;
+        int nSyncHeight = state->pindexBestKnownBlock ? state->pindexBestKnownBlock->nHeight : -1;
+
+        // Skip if node is in IBD
+        if ((nCommonHeight < chainActive.Tip()->nHeight - 10) && (nSyncHeight < chainActive.Tip()->nHeight - 10))
+            continue;
```

Comment:
maybe switch the 10 for DEFAULT_BLOCKS_FROM_TIP which we used in IsChainNearlySyncd(). I think this would be more in alignment with other areas of the code.